lib/web/apps.go

```diff
@@ -66,15 +66,58 @@ type CreateAppSessionRequest struct {
 	ClusterName string `json:"cluster_name"`
 }
 
+type GetAppFQDNResponse struct {
+	// FQDN is application FQDN.
+	FQDN string `json:"fqdn"`
+}
+
 type CreateAppSessionResponse struct {
 	// CookieValue is the application session cookie value.
 	CookieValue string `json:"value"`
 	// FQDN is application FQDN.
 	FQDN string `json:"fqdn"`
 }
 
+// getAppFQDN resolves the input params to a known application and returns
+// its valid FQDN.
+//
+// GET /v1/webapi/apps/:fqdn/:clusterName/:publicAddr
+func (h *Handler) getAppFQDN(w http.ResponseWriter, r *http.Request, p httprouter.Params, ctx *SessionContext) (interface{}, error) {
+	req := &AppResolveParams{
+		FQDNHint:    p.ByName("fqdnHint"),
+		ClusterName: p.ByName("clusterName"),
+		PublicAddr:  p.ByName("publicAddr"),
+	}
+
+	// Get an auth client connected with the user's identity.
+	authClient, err := ctx.GetClient()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get a reverse tunnel proxy aware of the user's permissions.
+	proxy, err := h.ProxyWithRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Use the information the caller provided to attempt to resolve to an
+	// application running within either the root or leaf cluster.
+	result, err := h.resolveApp(r.Context(), authClient, proxy, req)
+	if err != nil {
+		return nil, trace.Wrap(err, "Unable to resolve FQDN: %v", req.FQDNHint)
```

Comment:
Nit: We don't usually capitalize errors for better embedding as per go code review guide so I'd just say "**u**nable to...".